VexFlow 4.0 (starting with minimal migration)

rvilarl opened the pull request:
After failing to find the bugs on the previous migrations I decided to start a minimum change migration.

I went for:
- `import * as VF from "vexflow"` in order to have VF. in front of all the classes.
- the update of webpack in order to minimise the changes VF.class could not be used as constructor for example.

I will introduce further changes little by little making sure that the demo is not broken. With that the tie and cursor problems went away.

rvilarl commented:
@sschmidTU this one has not the tie and the cursor issues, ready to review. I had to update the packages though to achieve the minimal migration. There are quite a few things that can now be improved basically removal of any and casting. But I would like to get your feedback first in the current cut.

@ronyeh this one has now not known bugs and is using `import * as VF from "vexflow"`to have a clear identification of VexFlow classes.

sschmidTU commented:
@rvilarl Great, good to see that there's a working Vexflow 4 integration even if it doesn't have all the refactoring!
I'll definitely run the visual regression tests on this, will be very interesting exactly how differently Vexflow 4 renders.

rvilarl commented:
I have identified though another issue. Clef changes are missing! :(

**develop branch:**
![Captura de pantalla de 2021-11-13 17-30-21](https://user-images.githubusercontent.com/22865285/141651424-17b28dee-bcff-41f8-8885-b6dc31faab8f.png)

**this branch**
![Captura de pantalla de 2021-11-15 20-03-08](https://user-images.githubusercontent.com/22865285/141839417-c82191ce-fb92-47de-84dd-5b4cd507d847.png)

rvilarl commented:
I had to fix problems with fonts and canvas

sschmidTU commented:
At the moment i can't run the visual regression tests on this branch, because ̀ require` doesn't work anymore in the node script (̀ generateImages_browserless.js`). It wants everything to be ES modules and use import instead of require. I guess i'll have to port it to be an ES module.

sschmidTU commented:
Let me try to just take the build from this branch and use it on develop with the old packages and script.

sschmidTU commented:
Not sure about the in-measure clefs, did you change something there in Vexflow 4? It doesn't look to me like the API changed.

sschmidTU commented:
Ok, the new build doesn't work with `generateImages_browserless.js` even on develop.
```console
UnhandledPromiseRejectionWarning: ReferenceError: self is not defined
    at Object.<anonymous> (/Users/simon/OSMD/opensheetmusicdisplay-develop/build/opensheetmusicdisplay.min.js:2:224)
 ```
in the non-minified build (`npm run build:webpack-dev`), it's line 10, and it does look strange, `self` is indeed not defined beforehand.

By the way, what's also a little unfortunate is that the new build is 1.6MB, compared to 1.1MB with Vexflow 1.2.93. Maybe the unneeded music fonts aren't yet removed from Vexflow?

ronyeh commented:
> By the way, what's also a little unfortunate is that the new build is 1.6MB, compared to 1.1MB with Vexflow 1.2.93. Maybe the unneeded music fonts aren't yet removed from Vexflow?

Which font do you want to include in OSMD? If you are building straight from TypeScript, I can show you an example of how to include only a single font.

sschmidTU commented:
We want Gonville, at least for the first regression tests. Used to be the default (and only font i guess) in Vexflow 1.2.93.

ronyeh commented:
Got it. I'll update my typescript example to demonstrate bundling just Gonville and will send it to you shortly.

I haven't been following the details of this thread. What's the "self" error you are seeing? We have been careful to avoid directly referencing globals inside VexFlow 4, but if you can give me more info I can try to make sure it's not our fault :-). Or I can maybe help figure out the bug, because recently I've been dealing with a lot of browser vs node vs ESM vs CJS issues, and the global object is slightly different between them all.

sschmidTU commented:
@ronyeh The error is not in the main OSMD build, but in a node script i wrote to batch-generate PNG images (`generateImages_browserless.js`), which was basically a plain node script with `require()` and without a class/module kind of structure. I guess i'll have to port it to be a module, because `require` doesn't work in it anymore.
Thanks for offering help by the way :)

ronyeh commented:
Good luck! Yeah I guess ESM is the future, but migrating over has been a pain. Plus, you sort of have to support CJS/require forever anyways, until all the kids growing up now learn ESM only, :-D

ronyeh commented:
Hi, I put up a small example typescript project that you can clone and test out:

https://github.com/ronyeh/vexflow-example-typescript

```
npm install

npm run esbuildBundle

  build/app.js  1.3mb ⚠️

npm run esbuildSplit

  build/app.js                      520.5kb
  build/fonts/petaluma-NEVG2XA4.js  382.9kb
  build/fonts/bravura-IATPUOMW.js   204.9kb
  build/fonts/gonville-KIRL7A3V.js  145.3kb
  build/fonts/custom-X2HCXTBZ.js     13.5kb


npx http-server
```

Browse to  http://127.0.0.1:8080/app-module.html

Initial load is: 534 kB

After calling `await Flow.setMusicFont('Gonville');`, another 149 kB is loaded.


If you go to the tsconfig.json and comment out:
```
  // "vexflow": ["node_modules/vexflow/src/entry/vexflow-core.ts"]
```

And then use this line instead:
```
  "vexflow": ["node_modules/vexflow/src/entry/vexflow-core-with-gonville.ts"]
```

```
npm run esbuildSplit
```


Then load the app-module.html page again. It will load both chunks upon webpage load, for a total of 683 kB.

ronyeh commented:
I haven't gotten my example to do code splitting properly with webpack yet. So right now it uses esbuild, which is a very simple tool which works great for my small example.

I think part of the problem is that VexFlow 4 beta doesn't have proper/native ESM support yet. Once I get that into **VexFlow 4 beta 2**, I hope/think the webpack code splitting will work better.

For TypeScript projects, the magic is in these lines:

https://github.com/ronyeh/vexflow-example-typescript/blob/main/tsconfig.json#L2-L21

The tsconfig.json paths field allows you to map a module name to a specific typescript entry file. If you use:

```
"vexflow": ["node_modules/vexflow/src/entry/vexflow-core-with-gonville.ts"]
```

Then when you `import Vex from "vexflow"`, it will bundle Gonville by default, and allow for the other two fonts to be loaded via dynamic import. This only works when we enable code splitting in webpack / esbuild.

WARNING: There is a possibility that the entry file path may change in Beta 2. I might shorten the path to remove the src/ from the entry files:

```
vexflow/src/entry/vexflow-core-with-gonville.ts
==>
vexflow/entry/vexflow-core-with-gonville.ts
```

rvilarl commented on src/MusicalScore/Graphical/MusicSheetDrawer.ts:
```diff
@@ -45,7 +45,7 @@ export abstract class MusicSheetDrawer {
     public splitScreenLineColor: number;
     public midiPlaybackAvailable: boolean;
-    public drawableBoundingBoxElement: string = process.env.DRAW_BOUNDING_BOX_ELEMENT;
+    public drawableBoundingBoxElement: string = "None"; // process.env.DRAW_BOUNDING_BOX_ELEMENT;
```
This is the only one change that I had to do to avoid global variable access to get the tests running.

rvilarl commented:
@sschmidTU did you progress on running the tests?

sschmidTU commented:
@rvilarl by the way, as i said, we need to migrate the missing VexFlowPatch files that weren't already migrated to the vexflow repository (mostly in 3.x). So, just removing VexFlowPatch is not enough and will lead to regressions.

rvilarl commented:
@sschmidTU I know but copying .js files in the middle of the .ts files has no sense. We will have to run the visual regression testing and decide what patching has still to be ported to typescript.

sschmidTU commented:
@rvilarl that's why i said we need to migrate the files, meaning from .js to .ts and Vexflow 4.

sschmidTU commented:
@rvilarl By the way, I will have a look at all this at some point, we just had some other important priorities lately. Because we also need to migrate the scripts and VexFlowPatch files, there's a lot of work that needs to be done on this migration without a clear short term payoff. But I hope I can get to it soon.

sschmidTU commented:
@rvilarl Also, I can't `gh pr checkout 1094` currently, because it's non-fast-forward. Did you change anything recently that could cause this?
Also, if I checkout your fork and checkout the `migration/VF4.0` branch and run `npm install`, I get an error: `npm ERR! Cannot read property 'resolve' of undefined`. It used to work on earlier PR states.

rvilarl commented:
@sschmidTU I had to rebase due to new changes in develop that caused conflicts. You have to force a pull.

rvilarl commented:
It is still not possible to get the visual regresion working. When I run `npm run test:visual:build` I get the same error that I get when I try to use the library with my app pianoplay:
```
TypeError: Cannot read properties of undefined (reading 'Flow')
    at /home/rodrigo/git/opensheetmusicdisplay/build/opensheetmusicdisplay.min.js:2:1337974
    at /home/rodrigo/git/opensheetmusicdisplay/build/opensheetmusicdisplay.min.js:2:1644582
    at /home/rodrigo/git/opensheetmusicdisplay/build/opensheetmusicdisplay.min.js:2:1644588
    at /home/rodrigo/git/opensheetmusicdisplay/build/opensheetmusicdisplay.min.js:2:81
    at Object.<anonymous> (/home/rodrigo/git/opensheetmusicdisplay/build/opensheetmusicdisplay.min.js:2:223)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:190:29)
```
However `npm run build`, `npm test` and `npm start` run without issues.

@ronyeh perhaps you could help us here.

sschmidTU commented:
@rvilarl Try using a non-minified build to see where the exception occurs:
`npm run build:webpack-dev`
It looks like it's still using `Vex.Flow` somewhere, didn't you want to remove that everywhere?

rvilarl commented:
@sschmidTU yes I have checked that and `Vex.Flow` is nowhere in the sources but it is indeed in `opensheetmusicdisplay.min.js`:

```
Jt=globalThis.Vex
te=Jt.Flow
{Accidental:ee,Annotation:ie,Articulation:se,BarNote:ne,Beam:le,Bend:ae,BoundingBox:re,BoundingBoxComputation:oe,ChordSymbol:he,Clef:ce,ClefNote:be,Crescendo:me,Curve:ue,Dot:de,EasyScore:fe,Element:pe,Factory:ge,Font:xe,Formatter:Se,Fraction:ye,FretHandFinger:_e,GhostNote:ve,Glyph:we,GlyphNote:Ee,GraceNote:Te,GraceNoteGroup:Me,GraceTabNote:Pe,KeyManager:ze,KeySignature:Ae,KeySigNote:Le,Modifier:Ce,ModifierContext:Ne,MultiMeasureRest:De,Music:Re,Note:Be,NoteHead:Ie,NoteSubGroup:ke,Ornament:Fe,Parser:Oe,PedalMarking:Ve,Registry:Ue,RenderContext:We,Renderer:Ge,RepeatNote:He,Stave:Xe,Barline:je,StaveConnector:Ye,StaveHairpin:Ke,StaveLine:qe,StaveModifier:$e,StaveNote:Qe,Repetition:Ze,StaveTempo:Je,StaveText:ti,StaveTie:ei,Volta:ii,Stem:si,StringNumber:ni,Stroke:li,System:ai,Tables:ri,TabNote:oi,TabSlide:hi,TabStave:ci,TabTie:bi,TextBracket:mi,TextDynamics:ui,TextFormatter:di,TextNote:fi,TickContext:pi,TimeSignature:gi,TimeSigNote:xi,Tremolo:Si,Tuning:yi,Tuplet:_i,Vibrato:vi,VibratoBracket:wi,Voice:Ei,BarlineType:Ti,ModifierPosition:Mi}=Jt.Flow
```

rvilarl commented:
> @sschmidTU yes I have checked that and `Vex.Flow` is nowhere in the sources but it is indeed in `opensheetmusicdisplay.min.js`:
> 
> ```
> Jt=globalThis.Vex
> te=Jt.Flow
> ```

Source of the proble found:

https://github.com/0xfe/vexflow/blob/8fba46239218f1e8c13973e8357332899d8a5c4c/Gruntfile.js#L227-L233

@ronyeh I think that this the problem

rvilarl commented:
@sschmidTU I could run the regression test with the branch ESM from VexFlow. I included the diff files temporary in this PR so that you can have a look at them.

sschmidTU commented:
Please rather upload the diff files here instead of committing them to the branch, that just muddies the PR and diffs.

rvilarl commented:
It is for whatever reason not possible. I get this funny message "is not included in the list". Please take it from the PR and I will remove the commit from the branch.

![Captura de pantalla de 2021-11-29 14-54-07](https://user-images.githubusercontent.com/22865285/143880550-53f5c48b-17ae-4bbc-873b-0d8667278af8.png)

sschmidTU commented:
Thanks, I got it now, you can delete it. 32MB is not great for a git pull ;)

rvilarl commented:
@sschmidTU done

rvilarl commented:
@sschmidTU any feedback from the diffs?

sschmidTU commented on src/MusicalScore/Graphical/VexFlow/CanvasVexFlowBackend.ts:
```diff
@@ -47,8 +47,9 @@ export class CanvasVexFlowBackend extends VexFlowBackend {
-        this.renderer = new Vex.Flow.Renderer(this.canvas, this.getVexflowBackendType());
-        this.ctx = <Vex.Flow.CanvasContext>this.renderer.getContext();
+        // @ts-expect-error
+        this.renderer = new VF.Renderer(this.canvas, this.getVexflowBackendType());
```
we need `this.canvas as HTMLCanvasElement` here. I just tried npm start and it errors without this change. After this change, npm start works and creates the demo correctly.

rvilarl replied:
I know about this one, it is strange. It fails locally but not in the CI.

sschmidTU replied:
Ah, I see. Still, I see no problem with the casting (`as`), and otherwise npm start won't produce the demo.

rvilarl replied:
Done

rvilarl commented:
@sschmidTU I found the bug related to the clefs. The category strings were inconsistent and we renamed them. I have fixed the problem in the last commit and the clefs are now there. :)
![Captura de pantalla de 2021-12-10 16-51-56](https://user-images.githubusercontent.com/22865285/145602672-c247efde-036d-4d46-a69f-f2e738226a80.png)

rvilarl commented:
@sschmidTU I have updated vexflow with a prebuild hosted by @ronyeh. With this one the visual regression does work. It is time to consider if any of the vexflowpatches are still required. :)

sschmidTU commented:
@rvilarl Great! Are you sure there are no major visual regressions anymore?
Some of the vexflowpatches will definitely still be required, because they add some custom options. I guess we should have test samples for those.
Also, when there are no more regressions, besides verifying the VexFlowPatch changes also don't have regressions anymore, we still have to make sure that the build does not grow to 1.6MB by default before we can merge.
Do we need to use the Gonville-only Vexflow build for that? Is that also on npm?
If it's still a bit more complex to merge this due to these factors, we could also merge to a feature branch first, then merge to develop once we verified all issues are solved.

rvilarl commented:
@sschmidTU it is complex to say if there is something important missing on the visual regressions. All the files differ one way or the other. The size is now 1,4M. I will work with @ronyeh to get it compiled with Gonville only. Could you checkout this branch and have a look at the regressions I am sure that it is easier for you to identify the problems if any. 
A feature branch is also a possibility.
We are almost there! I was so happy to identify the Clef problem :)

sschmidTU commented:
Yes, good work! I'll take a look at the regressions.

ronyeh commented:
I can take a look. On my projects, I am able to get the bundle size down with only gonville pre-compiled. It's basically a webpack config. We need to somehow direct webpack to the correct entry file, and turn on code splitting.

rvilarl commented:
@sschmidTU do you use any other font than Gonville in the testing? Bravura for example?

sschmidTU commented:
@rvilarl No. Vexflow 1.2.93 is Gonville only i think.

rvilarl commented:
@sschmidTU it seems that OSDM is using somewhere the glyph `noteheadSquareWhite` which is not available in Gonville. I do not see where. Can you help to identify where this is used?

sschmidTU commented:
@rvilarl `noteheadSquareWhite ` is not in the OSMD source code, i don't know how i would go about finding where it's used indirectly.

rvilarl commented:
Ok thanks we will investigate further.

rvilarl commented:
@sschmidTU now only with "Gonville", **1083611 bytes**. The problem was  SetFont making reference to other fonts in:
- VexFlowMusicSheetCalculator.ts
- VexFlowMusicSheetDrawer.ts

There are also two types of note heads not supported by Gonville used in the test below (square & rectangle):
![OSMD_function_test_noteheadShapes musicxml_1](https://user-images.githubusercontent.com/22865285/145670643-dca55727-dfd2-4556-8d01-d368f4b41429.png)
![OSMD_function_test_noteheadShapes musicxml_1_Blessed](https://user-images.githubusercontent.com/22865285/145670645-825722c1-13cd-4b2e-80c4-f85aff2dd3b7.png)
![OSMD_function_test_noteheadShapes musicxml_1_Current](https://user-images.githubusercontent.com/22865285/145670646-375d04dc-525b-4bbd-b61f-43725c3d79fa.png)

sschmidTU commented:
@rvilarl nice! But weird, why did these notehead shapes work in 1.2.93 when they weren't in Gonville? that was the only font that 1.2.93 had.

rvilarl commented:
@sschmidTU I do not know how fonts were handled in 1.2.93 so I cannot answer the question, perhaps the set was enhanced manually like the NOGLYPHS. @0xfe can you help answering this question?
@sschmidTU would you move to Bravura? This is a maintained font and includes those glyphs.
If not, I could also include those glyphs manually in Gonville if @0xfe agrees.

sschmidTU commented:
Bravura will not be the default for the first version. The first version using Vexflow 4 should be as close to the one with Vexflow 1.2.93 as possible, otherwise it's not an upgrade but design change.

rvilarl commented:
@sschmidTU @0xfe @ronyeh I know why, those 'missing' glyphs are in custom_glyphs.ts. Custom should be included in all the vexflow-core-with-XXXX builds.

rvilarl commented:
@sschmidTU can you help me to resolve the problems with demo/plain_example.html. This example is based on your simplified example from the WIKI.

First it is failing to load due to "Uncaught SyntaxError: Invalid or unexpected token" I believe that it is due to blockFlöte & flöte in SubInstrement.ts because if I rename to blockFloete and floete, the error goes away.
Second after renaming I get a CORs error not allowing me to open the file "http://downloads2.makemusic.com/musicxml/MozaVeilSample.xml".

any tip on how to support html in plain_example.html?
would you agree to the renaming proposed above?

sschmidTU commented:
The rename is fine, though it's weird that it would create an error. Is UTF-8 not set correctly?
For the CORS error, you can't do much about that except set your browser to ignore CORS (which i think is not possible anymore in Chrome). This is not how you're supposed to load files on a local setup.
You have to either use a local server, or load the file via file dialog.
Here are some working examples, especially `indexSimpler.html`:
https://github.com/opensheetmusicdisplay/RawJavascript-usage-example

I'm not sure why you've added this "plain example" in the demo folder though,  this is all covered in our Wiki.
https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/wiki/Getting-Started

rvilarl commented:
> I'm not sure why you've added this "plain example" in the demo folder though, this is all covered in our Wiki. https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/wiki/Getting-Started

Well "plain example" is exactly the simplified example in the wiki page you refer to, I had expected it to work. I did put this example in a html in order to have something quick to check that our changes were working. No problem, we will get the migration working anyway. Just to check, it is good enough to get `npm run test:visual:build` and `npm start` runnig correctly, is not it.

sschmidTU commented:
Yes, that should be enough. and `npm test` ideally, but we can check that in the end, and it usually runs fine anyways.

rvilarl commented:
I have tried the WIP from @ronyeh https://github.com/0xfe/vexflow/pull/1232/commits/21af9b9f3ce879ba0a1007707694f797c36e17fb and everything works as expected. I am also very happy to see that the voices Gnossienne n 3 are now correctly aligned in my app Piano Play and the cursor is not going back and forward. :)

**vexflow4.0**

![Captura de pantalla de 2021-12-11 23-00-27](https://user-images.githubusercontent.com/22865285/145692965-daa9f52c-ebc0-48a8-b480-161aee80ff41.png)

**vexflow1.2.93**

![Captura de pantalla de 2021-12-11 22-52-21](https://user-images.githubusercontent.com/22865285/145692968-9c88995b-08d8-458a-bf77-606c2b555879.png)

rvilarl commented:
Notehead shapes are also fine

![OSMD_function_test_noteheadShapes musicxml_1](https://user-images.githubusercontent.com/22865285/145701267-537a9563-9ec7-48ca-bb0f-dfe2d3bbd7b8.png)

rvilarl commented:
@sschmidTU Custom font is now included, build size is 1094366. :)
Please have a look at the visual regression to see if we have to patch something in VexFlow 4

rvilarl commented:
@sschmidTU what about a new branch and the release of a 2.0.0-beta? So that me and other people can start playing with it?

sschmidTU commented:
I wouldn't make a beta release before VexFlowPatch isn't merged. You can always replace the build yourself if you want to use it, so I don't really see the point before that.

rvilarl commented:
> @sschmidTU Custom font is now included, build size is 1094366. :) Please have a look at the visual regression to see if we have to patch something in VexFlow 4

@sschmidTU did you look into VexflowPatch. I mean what patches are still required and if there are associated tests.

No beta ok, but what about a branch?This would allow that other contributors might help me. @ronyeh already made PRs to my fork but if we do it in a branch we would get more visibility.

sschmidTU commented:
A feature barnch would definitely be good, yes! If you want, I can already merge this to `feat/vexflow4`.

rvilarl commented:
> A feature barnch would definitely be good, yes! If you want, I can already merge this to `feat/vexflow4`.

yes please! I just rebased your latest commit.
 
> For VexflowPatch you just need to look into the readme.txt to see what's not merged to Vexflow yet.

ok

sschmidTU commented on src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts:
```diff
@@ -72,9 +72,9 @@ export class VexFlowMusicSheetCalculator extends MusicSheetCalculator {
+      VF.Flow.setMusicFont("Gonville", "Custom");
     } else if (this.rules.DefaultVexFlowNoteFont?.toLowerCase() === "petaluma") {
-      (Vex.Flow as any).DEFAULT_FONT_STACK = [(Vex.Flow as any).Fonts?.Petaluma, (Vex.Flow as any).Fonts?.Gonville, (Vex.Flow as any).Fonts?.Bravura];
+      VF.Flow.setMusicFont("Gonville", "Custom");
```
@rvilarl This should be Petaluma, right?

rvilarl replied:
No, you wanted to support Gonville only. This would increase the size of the build which is 1094366.

sschmidTU replied:
Well, for now, for the default build. But once we make a build with all fonts or with only petaluma maybe, this will be supported then, and until then this doesn't hurt, does it? Nobody will try to use Petaluma in a Gonville-only OSMD build.

rvilarl commented:
@sschmidTU I just rebased to include your latest commit. I found a bug also in package.json when I created the tgz file for my App. We are making some progress with the patches. Please look into the VexFlow patches in vexflow.

rvilarl commented:
@sschmidTU would you still consider to make a feature branch? It feels that getting in the develop branch will take ages :(

sschmidTU commented:
@rvilarl I don't really see the point before Vexflowpatch works in Vexflow4, but I can merge to a feature branch if that helps you.

rvilarl commented:
The only benefit would be that it would look more 'official' so perhaps someone could collaborate. I cannot imagine anyone making PRs to my fork. But you are right, I just looked and there are a lot of branches. I cannot imageine that someone looks there. Thanks anyway.

sschmidTU commented:
Well, it would also make it easier to check out. On the other hand, it would basically close the PR by merging, though you can always make further ones. I'll think about it.

rvilarl commented:
Of course I would keep on doing PRs on the branch until it was merged in develop. This PR will eventually get too big.
By the way I have seen that the Brackets (the ones joining staves) are not drawn. Give me a tip if you see something strange. And also if you see other problems on the visual differences.

rvilarl commented:
@sschmidTU here is the prerelease included, see last commit

rvilarl commented:
@sschmidTU I would say that the visual differences now are acceptable. I think that it would make sense to merge. What would be your concern? Please look at the visual differences. I had to remove one vexflow test that was crashing.

sschmidTU commented:
Cool! I'll take a look at the diffs as soon as i can.
The concerns are still the visual diffs and the not yet fully merged VexflowPatches.
Also, I added a few more recently like giving clefs and key signatures SVG groups, which would be unfortunate to see gone from OSMD again.

rvilarl commented:
You could raise PRs in VexFlow for the new changes. 
Are you really planning to move to VexFlow 4? In ten days I will be already 3 months doind rebases of this PR.

sschmidTU commented:
Yes, it would be great to do the Vexflow PRs when i have the time.
I'd love to move to Vexflow 4 as well, but not when it has no immediate benefit (for us) and it immediately brings multiple visual and functional regressions.

rvilarl commented:
Ok, just let me know when you are up to it. I will help you.

sschmidTU commented:
Ok, I've merged the last state of the integration to the branch `feat/vexflow4`, so others can test it and maybe build on it. Also makes it easier than checking out your branch, because rebasing always creates merge conflicts when pulling. Hopefully we'll soon get to merge the remaining Vexflow changes into vexflow.
